### PR TITLE
NIFI-8283 Value handling in ScanAccumulo processor

### PR DESCRIPTION
**Pull request by timeabarna**
Added value to record schema in order to displaying it in the scan result.
https://issues.apache.org/jira/browse/NIFI-8283

#### Description of PR

Added value to record schema in order to displaying it in the scan result.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java [hunk 1]
```diff
@@ -254,7 +258,12 @@ public void process(final InputStream in, final OutputStream out) throws IOExcep
 
                             try{
                                 final RecordSchema writeSchema = writerFactory.getSchema(flowAttributes, new KeySchema());
-                                try (final RecordSetWriter writer = writerFactory.createWriter(getLogger(), writeSchema, out)) {
+                                List<RecordField> fieldList = new ArrayList<>();
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java:261]: Recommend declaring this variable and other variables as `final` following the general pattern of the Processor.  Renaming the variable to something like `writeSchemaFields` or `recordSchemaFields` would also be helpful.

File: nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java [hunk 2]
```diff
@@ -270,8 +279,9 @@ public void process(final InputStream in, final OutputStream out) throws IOExcep
                                         data.put("columnQualifier", key.getColumnQualifier().toString());
                                         data.put("columnVisibility", key.getColumnVisibility().toString());
                                         data.put("timestamp", key.getTimestamp());
+                                        data.put("value", kv.getValue());
```
Comment (markap14) [inline nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java:282]: The schema here indicates that this field is to be a `String` type. That means that the value added into the map needs to be a String also, but it's of type `Value`. Need to ensure that it's properly converted to a `String` before putting into the Map.

File: nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java [hunk 3]
```diff
@@ -254,7 +258,12 @@ public void process(final InputStream in, final OutputStream out) throws IOExcep
 
                             try{
                                 final RecordSchema writeSchema = writerFactory.getSchema(flowAttributes, new KeySchema());
-                                try (final RecordSetWriter writer = writerFactory.createWriter(getLogger(), writeSchema, out)) {
+                                List<RecordField> fieldList = new ArrayList<>();
+                                fieldList.addAll(writeSchema.getFields());
+                                fieldList.add(new RecordField("value", RecordFieldType.STRING.getDataType()));
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java:263]: The Accumulo `Value` object contains a byte array which can be converted to UTF-8 String, but does conversion to String cover all use cases?

File: nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java (hunk 2 above)
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java:282]: As mentioned in the RecordField definition, `Value.toString()` converts the underlying byte array to a UTF-8 encoded String.  If that covers all use cases, this should be changed to `kv.getValue().toString()`.  Alternatively, if the underlying byte array should be represented, this should be changed to `kv.getValue().get()` and the RecordFieldType should be changed,

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
As @markap14 mentioned, the handling of `Value` should be corrected to ensure the expected output in the record.  Otherwise it looks like a straightforward adjustment.

File: nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java (hunk 2 above)
Comment (timeabarna) [reply nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java]: Thanks Mark for your help, modified to kv.getValue().toString()

File: nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java (hunk 3 above)
Comment (timeabarna) [reply nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java]: Thanks for your help exceptionfactory, String will cover most cases

File: nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java (hunk 1 above)
Comment (timeabarna) [reply nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java]: renamed variable to recordSchemaFields

File: nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java [hunk 4]
```diff
@@ -270,8 +281,9 @@ public void process(final InputStream in, final OutputStream out) throws IOExcep
                                         data.put("columnQualifier", key.getColumnQualifier().toString());
                                         data.put("columnVisibility", key.getColumnVisibility().toString());
                                         data.put("timestamp", key.getTimestamp());
+                                        data.put("value", kv.getValue().toString());
```
Comment (markap14) [inline nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java:284]: Is it possible here for the value to be null? I don't know much about accumulo, but i would guess so. If so, we need to check if null and if so avoid calling `toString()` on the null reference

Comment (timeabarna) [reply nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java]: To be on the safe side I've added the null check

**Comment by phrocker** [comment]
Hi @timeabarna  May want to add this as an option that is true by default. Thanks for the change.

There are cases where people do not want values and want to only get the keys.

**Comment by timeabarna** [comment]
@phrocker Thanks for your recommendation, I've updated the code accordingly

File: nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java [hunk 5]
```diff
@@ -143,6 +150,16 @@
             .addValidator(Validator.VALID)
             .build();
 
+    static final PropertyDescriptor VALUE_INCLUDED_IN_RESULT = new PropertyDescriptor.Builder()
+            .displayName("Value Included in Result")
+            .name("accumulo-value-inclusive")
+            .description("Beside keys and their values, accumulo value field will also be included in the result as UTF-8 Encoded String.")
+            .required(false)
+            .defaultValue("True")
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java:158]:
```suggestion
            .defaultValue("true")
```

File: nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java [hunk 6]
```diff
@@ -143,6 +150,16 @@
...
+            .allowableValues("True", "False")
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java:159]:
```suggestion
            .allowableValues("true", "false")
```

**Review by pvillard31 — CHANGES_REQUESTED** [review]
Just a minor comment to be consistent with other parts of the codebase. Otherwise LGTM.

File: nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java (hunk 6 above)
Comment (timeabarna) [reply nifi-nar-bundles/nifi-accumulo-bundle/nifi-accumulo-processors/src/main/java/org/apache/nifi/accumulo/processors/ScanAccumulo.java:159]: @pvillard31 thanks for your comment however within this processor all boolean type properties start with capital letters, so I followed the pattern to stay consistent.

**Review by pvillard31 — APPROVED** [review]

**Comment by pvillard31** [comment]
Merged to main, thanks @timeabarna

**Pull request closed without merging**